File: docs/04-Trees-and-BST/java/BST.java
public class BST {
    private static class TreeNode {
        int val;
        TreeNode left, right;
        
        TreeNode(int value) {
            this.val = value;
            this.left = this.right = null;
        }
    }

    private TreeNode root;

    public BST() {
        this.root = null;
    }

    public void insert(int val) {
        root = insertRec(root, val);
    }

    private TreeNode insertRec(TreeNode node, int val) {
        if (node == null) {
            return new TreeNode(val);
        }
        if (val < node.val) {
            node.left = insertRec(node.left, val);
        } else if (val > node.val) {
            node.right = insertRec(node.right, val);
        }
        return node;
    }

    public boolean search(int val) {
        return searchRec(root, val);
    }

    private boolean searchRec(TreeNode node, int val) {
        if (node == null) return false;
        if (node.val == val) return true;
        if (val < node.val) return searchRec(node.left, val);
        return searchRec(node.right, val);
    }

    public void inorder() {
        inorderRec(root);
        System.out.println();
    }

    private void inorderRec(TreeNode node) {
        if (node != null) {
            inorderRec(node.left);
            System.out.print(node.val + " ");
            inorderRec(node.right);
        }
    }

    public static void main(String[] args) {
        BST tree = new BST();
        System.out.println("Inserting 50, 30, 20, 40, 70, 60, 80 into BST:");
        tree.insert(50);
        tree.insert(30);
        tree.insert(20);
        tree.insert(40);
        tree.insert(70);
        tree.insert(60);
        tree.insert(80);
        
        System.out.print("In-order traversal: ");
        tree.inorder();
        
        System.out.println("Searching for 60: " + (tree.search(60) ? "Found" : "Not Found"));
        System.out.println("Searching for 90: " + (tree.search(90) ? "Found" : "Not Found"));
    }
}

File: docs/04-Trees-and-BST/js/bst.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class BST {
    constructor() {
        this.root = null;
    }

    insert(val) {
        this.root = this._insertRec(this.root, val);
    }

    _insertRec(node, val) {
        if (!node) {
            return new TreeNode(val);
        }
        if (val < node.val) {
            node.left = this._insertRec(node.left, val);
        } else if (val > node.val) {
            node.right = this._insertRec(node.right, val);
        }
        return node;
    }

    search(val) {
        return this._searchRec(this.root, val);
    }

    _searchRec(node, val) {
        if (!node) return false;
        if (node.val === val) return true;
        if (val < node.val) return this._searchRec(node.left, val);
        return this._searchRec(node.right, val);
    }

    inorder() {
        const result = [];
        this._inorderRec(this.root, result);
        console.log(result.join(" "));
    }

    _inorderRec(node, result) {
        if (node) {
            this._inorderRec(node.left, result);
            result.push(node.val);
            this._inorderRec(node.right, result);
        }
    }
}

// Driver testing
const tree = new BST();
console.log("Inserting 50, 30, 20, 40, 70, 60, 80 into BST:");
tree.insert(50);
tree.insert(30);
tree.insert(20);
tree.insert(40);
tree.insert(70);
tree.insert(60);
tree.insert(80);

console.print = console.log;
console.print("In-order traversal: ");
tree.inorder();

console.log("Searching for 60:", tree.search(60));
console.log("Searching for 90:", tree.search(90));
